barretenberg/cpp/src/barretenberg/vm/avm_trace/AvmMini_execution.cpp

```diff
@@ -0,0 +1,227 @@
+    return prover.construct_proof();
+}
+
+/**
+ * @brief Parsing of the supplied bytecode into a vector of instructions. It essentially
+ *        checks that each opcode value is in the defined range and extracts the operands
+ *        for each opcode.
+ *
+ * @param bytecode The bytecode to be parsed as a vector of bytes/uint8_t
+ * @throws runtime_error exception when the bytecode is invalid.
+ * @return Vector of instructions
+ */
+std::vector<Instruction> Execution::parse(std::vector<uint8_t> const& bytecode)
+{
+    std::vector<Instruction> instructions;
+    size_t pos = 0;
+    const auto length = bytecode.size();
+
+    static_assert(sizeof(uint32_t) / sizeof(uint8_t) == AVM_OPERAND_BYTE_LENGTH);
+
+    while (pos < length) {
+        const uint8_t opcode_byte = bytecode.at(pos);
+        pos += AVM_OPCODE_BYTE_LENGTH;
+
+        if (!Bytecode::is_valid(opcode_byte)) {
+            throw std::runtime_error("Invalid opcode byte: " + std::to_string(opcode_byte));
+        }
+
+        const auto opcode = static_cast<OpCode>(opcode_byte);
+        auto in_tag_u8 = static_cast<uint8_t>(AvmMemoryTag::U0);
+
+        if (Bytecode::has_in_tag(opcode)) {
+            if (pos + AVM_IN_TAG_BYTE_LENGTH > length) {
+                throw std::runtime_error("Instruction tag missing at position " + std::to_string(pos));
+            }
+            in_tag_u8 = bytecode.at(pos);
+            if (in_tag_u8 == static_cast<uint8_t>(AvmMemoryTag::U0) || in_tag_u8 > MAX_MEM_TAG) {
+                throw std::runtime_error("Instruction tag is invalid at position " + std::to_string(pos) +
+                                         " value: " + std::to_string(in_tag_u8));
+            }
+            pos += AVM_IN_TAG_BYTE_LENGTH;
+        }
+
+        auto const in_tag = static_cast<AvmMemoryTag>(in_tag_u8);
+        std::vector<uint32_t> operands{};
+        size_t num_of_operands{};
+        size_t operands_size{};
+
+        if (opcode == OpCode::SET) {
+            switch (in_tag) {
+            case AvmMemoryTag::U8:
+                num_of_operands = 2;
+                operands_size = 5;
+                break;
+            case AvmMemoryTag::U16:
+                num_of_operands = 2;
+                operands_size = 6;
+                break;
+            case AvmMemoryTag::U32:
+                num_of_operands = 2;
+                operands_size = 8;
+                break;
+            case AvmMemoryTag::U64:
+                num_of_operands = 3;
+                operands_size = 12;
+                break;
+            case AvmMemoryTag::U128:
+                num_of_operands = 5;
+                operands_size = 20;
+                break;
+            default:
+                throw std::runtime_error("Instruction tag for SET opcode is invalid at position " +
+                                         std::to_string(pos) + " value: " + std::to_string(in_tag_u8));
+                break;
+            }
```

Comment:
Maybe add a comment here just explaining this if-statement and the switch case for SET. Makes sense to me though.

Reply:
Added one.